### PR TITLE
TST: loc misbehaves when Period is at start of 3-level MultiIndex

### PR DESCRIPTION
**Pull request by proost**
If index is MultiIndex and level of 0 is PeriodIndex, loc function raise
exception if all input of loc does not match index values

Background: This bug only happens when MultiIndex's level is 3 and first level index is PeriodIndex. In this situation, if someone access one row using a '.loc' with a miss match key, then would not raise exception.

Someone already change what i try to do `parse_time_string` function in '_libs.tslibs.parsing'.
in the past,
```
def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
    if not isinstance(arg, str):
        return arg
```

What i try to do:

```
def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
    if not isinstance(arg, str):
        raise TypeError
```

now in master:
```
def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
    if not isinstance(arg, str):
        raise TypeError("parse_time_string argument must be str")
```
Just add tests for issue. 

- [x] closes (#20684)
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Can you add tests specific to parse_time_string.  That would go in tests.tslibs.test_parsing

**Comment by minhoryang** [comment]
@jbrockmendel ping?

File: pandas/tests/indexes/period/test_period.py [hunk 1]
```diff
@@ -646,6 +646,15 @@ def test_insert(self):
             result = period_range("2017Q1", periods=4, freq="Q").insert(1, na)
             tm.assert_index_equal(result, expected)
 
+    def test_contains_raise_key_error(self):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_period.py]: the test name could be more informative

File: pandas/tests/tslibs/test_parsing.py [hunk 2]
```diff
@@ -209,3 +209,13 @@ def test_try_parse_dates():
 
     expected = np.array([parse(d, dayfirst=True) for d in arr])
     tm.assert_numpy_array_equal(result, expected)
+
+
```
Comment (jreback) [inline pandas/tests/tslibs/test_parsing.py]: test name could ben more informative

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -295,9 +295,10 @@ Missing
 MultiIndex
 ^^^^^^^^^^
 
+- Bug in :meth:`DataFrame.loc` where :class:`PeriodIndex` is in level of 0 raise exception if entire key is not matched (:issue:`20684`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: pls reword to be more clear

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.0.rst [hunk 4]
```diff
@@ -297,8 +297,9 @@ Missing
 MultiIndex
 ^^^^^^^^^^
 
+
 - Constructior for :class:`MultiIndex` verifies that the given ``sortorder`` is compatible with the actual ``lexsort_depth``  if ``verify_integrity`` parameter is ``True`` (the default) (:issue:`28735`)
--
+- If `MultiIndex` nlevel is 3 and first level index in `MultiIndex` is `PeriodIndex` , `.loc` could not raise exception when key is not matched (:issue:`20684`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this note is too specific, just say using .loc with a PeriodIndex as a level in a MultiIndex

File: pandas/tests/indexes/period/test_period.py [hunk 5]
```diff
@@ -646,6 +646,21 @@ def test_insert(self):
             result = period_range("2017Q1", periods=4, freq="Q").insert(1, na)
             tm.assert_index_equal(result, expected)
 
+    def test_contains_raise_error_if_first_level_index_is_period_index(self):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_period.py]: can you try  on 2nd level as well; I think the bug manifests on any level

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you also make sure we have tests for the entire example in the OP (IOW the working cases as well)

**Comment by proost** [comment]
@jreback  could you also check #29038 ?

File: pandas/tests/indexes/period/test_period.py [hunk 6]
```diff
@@ -649,6 +649,72 @@ def test_insert(self):
...
+        assert ybc.loc[("2017", "b1", "c1"), "v"] == 1
+
+        msg = r"'2017', 'b1', 'wobble'"
+        with pytest.raises(KeyError, match=msg):
+            ybc.loc[("2017", "b1", "wobble"), "v"]
+
+        pb = DataFrame({"p": [Period("2017")], "b": ["b1"], "v": [1]}).set_index(
+            ["p", "b"]
+        )
+        assert pb.loc[(Period("2017"), "b1"), "v"] == 1
```
Comment (jreback) [inline pandas/tests/indexes/period/test_period.py]: can you add some comments on groups of tests here. really hard to tell why things *should* fail as there are a lot of tests. is there any way to paramterize these? these seems duplicated (but hard to tell)

File: doc/source/whatsnew/v1.0.0.rst [hunk 7]
```diff
@@ -341,6 +341,7 @@ MultiIndex
 
 - Constructior for :class:`MultiIndex` verifies that the given ``sortorder`` is compatible with the actual ``lexsort_depth``  if ``verify_integrity`` parameter is ``True`` (the default) (:issue:`28735`)
 -
+- Using `.loc` with a `PeriodIndex` as a level in a `MultiIndex`, `.loc` could not raise exception when key is not matched (:issue:`20684`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: you didn't make any code changes, is this a test only PR?

Comment (proost) [reply doc/source/whatsnew/v1.0.0.rst]: @jreback 
Yes. Only add tests.
I try to change `PeriodIndex.get_loc` in two ways.
First, I try to remove `PeriodEngine.get_loc` in `PeriodIndex.get_loc`, so just compute `.get_loc` once, not several in try-except statement. I change several ways  But always stuck in test cases. Changing some test cases is inevitable. So I give it up this way.
Second, I try to change `PeriodEngine.get_loc` to accept `Period` object not only `Period.ordinal`. (Original `PeriodEngine.get_loc` only accept `Period.ordinal`) This way is success in that work successfully and pass all test cases. However Problem is most cases user don't use `Period` object using `PeriodIndex.get_loc` And `PeriodEngine.get_loc` can't accept "intolerance" argument. So Integrating `PeriodIndex.get_loc` can't be possible.
In conclusion, Refactoring `PeriodIndex.get_loc` fails. Someone who is good at programming more than me can do it maybe.

Comment (proost) [reply doc/source/whatsnew/v1.0.0.rst]: @jreback 
And while i get review, someone change what i try to change

Comment (jreback) [reply doc/source/whatsnew/v1.0.0.rst]: ok, then don't need the whatsnew note. where was the change, can you point to a PR?

**Review by jreback — CHANGES_REQUESTED** [review]
ok looks fine, if you'd remove the note, otherwise lgtm. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @proost

**Comment by proost** [comment]
@jbrockmendel  @jreback 
Thank you for your feedback.
